Add docs for exportPdf plugin

Dumluregn opened the pull request:
Two articles added: in `Features` and `Examples` sections.

In the other files - changes updating the statements like "we don't save to PDF" etc.

There are errors during building docs - those are non-existing yet links to API, docs work normally.

In the example plugin is not installed yet. I wanted docs to work normally, so this can be fixed after the package will be available on npm and in addons (or maybe there is another way to attach the plugin now I'm not aware of?).

f1ames commented on docs/features/exportpdf/README.md:
```diff
@@ -0,0 +1,104 @@
+
+Due to the differences between browsers and operating systems it is not always possible to reach a perfect match between content in editor and PDF. However a lot can be done to ensure the differences will be hardly noticable.
+
+For example, make the editor width correspond with the chosen paper format for output file - e.g. for the `A4` format (which is used by the export service [by default](https://pdf-converter.cke-cs.com/docs#section/PDF-options/Page-format)) the editor width should be equal to `840px`. Then mind the margins; if they are changed for the PDF using [custom config](https://pdf-converter.cke-cs.com/docs ...
```
I would skip `For example` since it is the most important thing here :stuck\_out\_tongue: Also section on margins should link to something like `addCss` so person reading it also can check how to adjust margins in CKEditor 4 itself.

f1ames commented on docs/features/exportpdf/README.md:
```diff
@@ -0,0 +1,104 @@
+
+### Custom CSS rules
+
+Besides the inline styles, in order to mimic the editor content all the CKEditor 4 CSS rules are sent to the service. But let's say your company has a strict heading color policy: it has to be <span style="color:#4B0082">indigo</span> (`#4B0082`), definitely not <span style="color:#483D8B">dark slate blue</span> (`#483D8B`)! Fortunately you do not have to remember to change it every time. It is p ...
```
`{@linkapi CKEDITOR.config.exportPdf_stylesheet CKEDITOR.config.exportPdf_stylesheet}`

That will require some magic, since the plugin is not the part of presets we will need to copy it to `ckeditor4` directory for the time of docs generation so the API docs could be parsed :thinking: And having in mind that plugin code will be obfuscated we will need to extract API docs to a separate file for plugin dist probably? :thinking: :scream:

Dumluregn replied:
Another option is putting all the config options in the `features` instead of in `API` section - AFAICS this is the way this problem was resolved for [Spreadsheet plugin](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_spreadsheets.html#automatically-convert-existing-tables)?

Dumluregn replied:
Needless to say it's not a dream way of dealing with this problem.

f1ames replied:
> Another option is putting all the config options in the `Features` instead of `API` section - AFAICS this is the way this problem was resolved for [Spreadsheet plugin](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_spreadsheets.html#automatically-convert-existing-tables)?

Yes, that was kind of a workaround there. When working on Export to PDF dist version building recently I already see that extracting comments is doable so we should be fine here :crossed\_fingers:

f1ames commented on docs/features/exportpdf/README.md:
```diff
@@ -0,0 +1,104 @@
+Plugin provides a custom {@linkapi CKEDITOR.editor#exportPdf exportPdf event}. It can be used for custom data processing (e.g. to ensure the output text will be black, not <span style="color:pink;background-color:yellow">pink on yellow background</span>). Editor uses it too, so just remember to add the right priority to the listener:
+
+* 1-14: Data is available in the original string format.
+* 15: Data is processed by the plugin.
```
```suggestion
* 15: Data is preprocessed by the plugin.
```

Can we elaborate here (or below) what will be changed in this step?

Dumluregn replied:
As we obfuscate the plugin I wasn't sure how mysterious we want to be, but since we have e.g. a section about relative and absolute URLs I guess it's not such a big secret :see\_no\_evil:

f1ames commented on docs/features/exportpdf/README.md:
```diff
@@ -0,0 +1,104 @@
+
+* 1-14: Data is available in the original string format.
+* 15: Data is processed by the plugin.
+* 16-19: Data that will be sent to the service can be modified.
```
Basically, data can be modified at any point so this is not entirely true - it just depends if one would like to modify original data or the one after our modifications.

f1ames commented on docs/features/exportpdf/README.md:
```diff
@@ -0,0 +1,104 @@
+
+It is even possible to run some asynchronous tasks before sending data to the server. To do this simply stop the event and set the flag telling that process is not finished yet. After it is done, remove flag and refire the `exportPdf` event:
+
+	editor.on( 'exportPdf', function( evt ) {
+		// Stop the event if the async process is still on.
+		if ( !evt.data.asyncDone ) {
+			evt.cancel();
+		} else {
+			delete evt.data.asyncDone;
+		}
+	} );
+
+	editor.on( 'exportPdf', function( evt ) {
+		// Here e.g. some AJAX request can be sent; after it is done set the flag to `true`:
+		evt.data.asyncDone = true;
+
+		// Refire the event.
+		editor.fire( 'exportPdf', evt.data );
+	}, null, null, 1 );
```
```js
    editor.on( 'exportPdf', function( evt ) {
        // Let's call some asynchronous function here, like an Ajax request. Flag processing as in progress.
        evt.data.asyncDone = false;

        ajaxRequestForAdditionalData( evt.editor, function() {
            // Ajax call is done, let's mark processing as done and refire 'exportPdf` event.
            evt.data.asyncDone = true;
            editor.fire( 'exportPdf', evt.data );
        } );
    }, null, null, 1 );

    editor.on( 'exportPdf', function( evt ) {
        // Stop the event if the async process is still on.
        if ( !evt.data.asyncDone ) {
            evt.cancel();
        }
    } );
```

WDYT?

f1ames commented on docs/features/exportpdf/README.md:
```diff
@@ -0,0 +1,104 @@
+
+### Relative vs absolute URLs
+
+Assets like images can be attached using relative URLs, but before data is sent to the service, relative links are converted to absolute ones. In some cases it will mean that data will not be accessible by the server (e.g. if it is referenced locally or through the infranet) - remember to expose such assets globally.
```
```suggestion
Assets like images can be attached using relative URLs, but before data is sent to the service, relative links are converted to absolute ones. In some cases it will mean that data will not be accessible by the server (e.g. if it is referenced locally or through the infranet) - remember to expose such assets publicly or use absolut URLs to publicly available assets.
```

Maybe we could also mention [`baseHref`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-baseHref). So one can have assets storage exposed somewhere else?

f1ames commented on docs/guide/dev/example_setups/README.md:
```diff
@@ -36,7 +36,7 @@ The Article Editor is based on the [Standard package](https://ckeditor.com/ckedi
 ## Document Editor
 
-The Document Editor demo showcases a more robust editor designed for creating documents which are usually later printed or exported to PDF files using tools like [wkhtmltopdf](https://github.com/wkhtmltopdf/wkhtmltopdf) or [PhantomJS](https://github.com/ariya/phantomjs) (note: PhantomJS 2.x currently has a [known zoom issue](https://github.com/ariya/phantomjs/issues/13997)).
+The Document Editor demo showcases a more robust editor designed for creating documents which can later be {@link features/exportpdf/README exported to PDF} using the official [Export to PDF](https://ckeditor.com/cke4/addon/exportpdf) plugin.
```
Good catch :+1: Have you checked entire docs for other things like this and places where we can link to PDF Export plugin too?

Dumluregn replied:
Yes. There is also one place we could also change the editor API docs: https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_plugins_print.html, but it would require a separate PR (as it's in `ckeditor4` repo). I can extract an issue if you want.

f1ames replied:
Could you extract it to separate issue?

Dumluregn replied:
ckeditor/ckeditor4#4077 - again I didn't know how mysterious should I be :see\_no\_evil: but after [this comment](https://github.com/ckeditor/ckeditor4/issues/4061#issuecomment-630714428) I guess my issue is OK.

f1ames commented on docs/sdk/examples/exportpdf.html:
```diff
@@ -0,0 +1,178 @@
+
+			<h2 style="text-align:center"><span style="font-family:Georgia,serif"><span style="color:#2980b9">The Flavorful Tuscany Meetup Schedule</span></span></h2>
+
+			<table border="1" cellspacing="0" style="border-collapse:collapse; width:597px">
```
I wonder if we can make the whole table centered? Would it work in generated PDF?

Dumluregn replied:
Yes, it works :+1: I also centered the content of table.

Dumluregn commented:
> 1.  Is `docs/assets/img/exportpdf_01.png` image HiDPI ready?

Yes :+1: 

> 1.  Missing icon in examples

As far as I understand the docs build process, this icon is linked to a plugin by `node_modules`, so after the plugin is released and added to project this icon should work (same as e.g. for [`removeformat`](https://github.com/ckeditor/ckeditor4-docs/blob/master/docs/sdk/examples/removeformat.html#L44)).

f1ames commented on docs/sdk/examples/exportpdf.html:
```diff
@@ -0,0 +1,179 @@
+		<script>
+			// CKEDITOR.plugins.addExternal( 'pdfexport', '../../../ckeditor4/{%CKEDITOR_VERSION%}/{%CKEDITOR_EXAMPLES_SLUG%}/assets/plugins/pdfexport/', 'plugin.js' );
+
+			if ( CKEDITOR.env.ie && !CKEDITOR.env.version < 11 ) {
```
There is something wrong with this condition, should it be:

```javascript
CKEDITOR.env.ie && CKEDITOR.env.version < 11
```

without (`!`)?

Dumluregn commented on docs/features/exportpdf/README.md:
```diff
@@ -0,0 +1,106 @@
+
+### Data preprocessing
+
+Plugin provides a custom {@linkapi CKEDITOR.editor#exportPdf exportPdf event}. It can be used for custom data processing (e.g. to ensure the output text will be black, not <span style="color:pink;background-color:yellow">pink on yellow background</span>). Editor uses it too, so just remember to add the right priority to the listener:
```
I also linked to a community plugin: https://ckeditor.com/cke4/addon/base64image.

f1ames replied:
That's a good idea, however... Basically, by mentioning community plugin we recommend it so we should make sure everything works fine there. Also there are [more base64 image plugins](https://ckeditor.com/cke4/addons/search/plugins/base64) so we should link to the most reliable one :thinking: Or we could link to list of all image base64 plugins (e.g. `https://ckeditor.com/cke4/addons/search/plugins/base64`).

Since we have issue for converting images to base64 representation already reported I'm not sure if this mention is needed?

Dumluregn commented on docs/features/exportpdf/README.md:
```diff
@@ -0,0 +1,106 @@
+
+### Custom CSS rules
+
+Besides the inline styles, in order to mimic the editor content, all the CKEditor 4 CSS rules are sent to the service. But let's say your company has a strict heading color policy: it has to be <span style="color:#4B0082">indigo</span> (`#4B0082`), definitely not <span style="color:#483D8B">dark slate blue</span> (`#483D8B`)! Fortunately you do not have to remember to change it every time. It is  ...
```
For classic editor I decided to mention [`config.contentsCss`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-contentsCss) method instead of `config.exportPdf_stylesheet`, as it has basically the same use cases but makes more sense as the changed styles will already be visible in the editor, not just in the PDF.

f1ames commented on docs/features/exporttopdf/README.md:
```diff
@@ -0,0 +1,117 @@
+
+#### Divarea and Inline editor
+
+Here, due to the nature of these editors (as they use the styles of the webpage they are embedded in instead of editor's own rules), you should attach the additional stylesheet (one or more) using {@linkapi CKEDITOR.config.exportPdf_stylesheet CKEDITOR.config.exportPdf_stylesheet} configuration option. In such case these are the only non-inline styles that are sent.
```
> as they use the styles of the webpage they are embedded in instead of editor's own rules

Well, these editors have it's own rules defined (same as classic editor), but they also inherit styles from parent page.

f1ames replied:
Maybe mention also that if styles are not attached only inline styles will be preserved and some defaults (similar ot browser default styling) applied?

Dumluregn replied:
Wasn't easy to avoid using the word `default` ("default browser styling is sent by default") :stuck\_out\_tongue\_closed\_eyes:

f1ames commented on docs/features/exporttopdf/README.md:
```diff
@@ -0,0 +1,117 @@
+
+Plugin provides a custom {@linkapi CKEDITOR.editor#exportPdf exportPdf event}. It can be used for custom data processing (e.g. to ensure the output text will be black, not <span style="color:pink;background-color:yellow">pink on yellow background</span>). Editor uses it too, so just remember to add the right priority to the {@linkapi CKEDITOR.editor.on event listener}:
+
+* 1-14: Data is available in the original string format (it is extracted using {@linkapi CKEDITOR.editor.getData CKEDITOR.editor.getData} method).
```
```suggestion
* 1-14: Data is available in the original string format (it is extracted using {@linkapi CKEDITOR.editor.getData `CKEDITOR.editor.getData()`} method).
```

:wink:

f1ames commented on docs/features/exporttopdf/README.md:
```diff
@@ -0,0 +1,117 @@
+Plugin provides a custom {@linkapi CKEDITOR.editor#exportPdf exportPdf event}. It can be used for custom data processing (e.g. to ensure the output text will be black, not <span style="color:pink;background-color:yellow">pink on yellow background</span>). Editor uses it too, so just remember to add the right priority to the {@linkapi CKEDITOR.editor.on event listener}:
+
+* 1-14: Data is available in the original string format (it is extracted using {@linkapi CKEDITOR.editor.getData CKEDITOR.editor.getData} method).
+* 15: Data is preprocessed by the plugin: image relative paths are changed to absolute ones and editor's content is wrapped into a container with appropriate classes for styling.
```
Do you think we could add some short info about the wrapper class name? So one could use correctly namespaced CSS rules which will be only applied for PDF file?

It won't fit here IMHO, but maybe somewhere in the Custom CSS section?

f1ames commented on docs/sdk/examples/exporttopdf.html:
```diff
@@ -0,0 +1,182 @@
+		</p>
+		<p>
+			It introduces a <span class="button_icon" data-icon="exportpdf" title="Export to PDF">&nbsp;</span><b>new toolbar button</b>.
+			When it's clicked, data from editor is sent to the service maintained by <a href="https://ckeditor.com/ckeditor-cloud-services/">CKEditor Cloud Services</a>.
+			The editor takes care of the received response and saves it as a file.
```
```suggestion
			When it's clicked, data from the editor is sent to the PDF service maintained by <a href="https://ckeditor.com/ckeditor-cloud-services/">CKEditor Cloud Services</a>.
			The PDF service transforms received HTML into a PDF file which is sent back to editor and saved as a PDF file.
```

WDYT?

Dumluregn replied:
Actually service transforms the HTML to a binary file :slightly\_smiling\_face: besides from that - :+1:

f1ames replied:
Well, basically every file is a binary file (and [text files are usually treated as separate category](https://www.nayuki.io/page/what-are-binary-and-text-files)) :wink: So it's both binary and a PDF file.

f1ames replied:
But I'm fine with current wording ofc:+1:

f1ames commented:
One more thing - I have noticed \`service\` is used few times in the docs and also \`endpoint\`. To not confuse readers maybe we should stick to one name - and probably it should be self explanatory, so something like \`PDF service\` / \`PDF endpoint\`?

f1ames commented:
One more thing I realized is that we are missing any info on plugin installation. Some people will come to docs from NPM package where it will be explained briefly, but if they visit the docs first it should be also mentioned there.

Basically, this is the similar situation as for Spreadsheets plugin so one can use NPM, Online Builder or manual approach so you could resue Spreadsheets guides to some extend - https://ckeditor.com/docs/ckeditor4/latest/guide/dev_spreadsheets.html#quick-start.

Dumluregn commented:
> I have noticed \`service\` is used few times in the docs and also \`endpoint\`

I think `service` is a better suited word for such a guide (`endpoint` being to technical), so I replaced it as `PDF service`.

I also added a mention that stylesheet links are also absolutised (in the same section as images).

f1ames commented on docs/features/exporttopdf/README.md:
````diff
@@ -0,0 +1,145 @@
+npm install ckeditor4-plugin-exportpdf
+```
+
+If CKEditor 4 was also installed through npm (using the `ckeditor4` package) in your project, the plugin will create a symbolic link itself in the CKEditor 4 `plugins/` directory so you do not need to move or copy any files.
````
It won't until we provide some postinstall script so this section is not valid ATM.

f1ames commented on docs/features/exporttopdf/README.md:
````diff
@@ -0,0 +1,145 @@
+If you prefer not to touch any files, you may use the [`CKEDITOR.plugins.addExternal()`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_plugins.html#method-addExternal) API instead to load the Export to PDF plugin directly from the `node_modules/` directory:
+
+```js
+CKEDITOR.plugins.addExternal( 'exportpdf', './node-modules/ckeditor4-plugin-exportpdf/exportpdf/' );
````
```suggestion
CKEDITOR.plugins.addExternal( 'exportpdf', './node-modules/ckeditor4-plugin-exportpdf/' );
```

This path is shorter :point\_up: In Spreadsheets plugin there was specific files structure but in this case we will have regular one so `plugin.js` file is in the root plugin/package directory.

Dumluregn replied:
I wasn't sure about it, thx :+1:

AnnaTomanek commented:
The editor in the sample should be configured to look like a page-like document, just like we have it in the [Paste from Word](https://ckeditor.com/docs/ckeditor4/latest/examples/pastefromword.html) sample or the [Document](https://ckeditor.com/ckeditor-4/demo/#document) demo.

In addition to that, for rendering to be close to 1:1, we should set the "page" to the dimensions of the A4 page.

You can check out the [CKEditor 5 sample](https://ckeditor.com/docs/ckeditor5/latest/features/export-pdf.html) for inspiration, too :wink:

Dumluregn commented:
Added the plugin to docs - sample is now live (and beautiful) :tada: And adjusted for best rendering.

AnnaTomanek commented:
Can we adjust the margins in the sample document? They seem to be unnaturally small... How about we set them to the values that a default Word document in A4 has?

AnnaTomanek commented:
I reviewed and corrected the guide, it looks nice :clap: 

I am missing a bit some more focus on the configuration of this feature - this is something that can generate plenty of customer care requests unless we document it correctly as people may be wondering why what they see in the PDF file does not look the same as their editor content.

What do you think about adding some more content about things like stylesheets and web fonts plus some examples? These two sections from CKE5 documentation could be an inspiration :wink: 

*   https://ckeditor.com/docs/ckeditor5/latest/features/export-pdf.html#html-to-pdf-converter-features
*   https://ckeditor.com/docs/ckeditor5/latest/features/export-pdf.html#examples

vokiel commented on docs/sdk/examples/exporttopdf.html:
```diff
@@ -0,0 +1,201 @@
+				height: 700,
+				extraPlugins: 'colorbutton,font,justify,print,tableresize,liststyle,pagebreak,exportpdf',
+				toolbar: [
+					{ name: 'document', items: [ 'exportPdf', 'PageBreak', 'Source' ] },
```
Typo:

```suggestion
					{ name: 'document', items: [ 'ExportPdf', 'PageBreak', 'Source' ] },
```

AnnaTomanek commented on docs/features/exporttopdf/README.md:
```diff
@@ -0,0 +1,213 @@
+Here are a few configuration tips that will make it easier to achieve a close to 1:1 rendering between the editor content and the output PDF document:
+
+* **Editor width.**
+  Make the editor width correspond with the chosen paper format for output file. For example, for the `A4` format (which is used by the PDF converter service [by default](https://pdf-converter.cke-cs.com/docs#section/PDF-options/Page-format)), the editor width should be equal to `805px`.
```
A4 is the default format of the converter, while Letter is the most popular US format. It would be cool to have the value for Letter here, too.

Dumluregn replied:
Added info. However... I just discovered that _something_ (don't know if it's macOS, Chrome or CKEditor) detects if you have a mouse plugged and then modifies the styles on page (scrollbar in particular), which makes it impossible to determine the width that will always work (see the gif below - I'm just plugging and unplugging mouse) :disappointed:  I assumed that more people use mouse than doesn't and both values (for `A4` and `Letter` ) are correct if mouse is connected.

![](https://user-images.githubusercontent.com/10883184/84155611-cfeda480-aa68-11ea-8cad-cc114928d0d5.gif)

AnnaTomanek replied:
:woman\_facepalming:  I don't even...

Thanks for researching this! Let's wait for feedback...

f1ames commented:
Rebased onto latest `stable.`
